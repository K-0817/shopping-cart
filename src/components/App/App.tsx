import { useEffect } from 'react';

import Loader from 'components/Loader';
import { GithubCorner, GithubStarButton } from 'components/Github';
import Filter from 'components/Filter';
import Products from 'components/Products';
import Cart from 'components/Cart';

import { useProducts } from 'contexts/products-context';

import * as S from './style';

function App() {
  const { isFetching, products, fetchProducts } = useProducts();

  useEffect(() => {
    fetchProducts();
  }, [fetchProducts]);

  return (
    <S.Container>
      {isFetching && <Loader />}
      <GithubCorner />
      <S.TwoColumnGrid>
        <S.Side>
          <Filter />
          <GithubStarButton />
        </S.Side>
        <S.Main>
          <S.MainHeader>
            <p>{products?.length} Product(s) found</p>
          </S.MainHeader>
          <Products products={products} />
        </S.Main>
      </S.TwoColumnGrid>
      <Cart />
    </S.Container>
  );
}

export default App;
