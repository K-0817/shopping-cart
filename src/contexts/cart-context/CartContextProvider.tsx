import { createContext, useContext, FC, useState } from 'react';
import { ICartProduct, ICartTotal } from 'models';

export interface ICartContext {
  isOpen: boolean;
  setIsOpen(state: boolean): void;
  products: ICartProduct[];
  setProducts(products: ICartProduct[]): void;
  total: ICartTotal;
  setTotal(products: any): void;
}

const CartContext = createContext<ICartContext | undefined>(undefined);
const useCartContext = (): ICartContext => {
  const context = useContext(CartContext);

  if (!context) {
    throw new Error('useCartContext must be used within a CartProvider');
  }

  return context;
};

const totalInitialValues = {
  productQuantity: 0,
  installments: 0,
  totalPrice: 0,
  currencyId: 'USD',
  currencyFormat: '$',
};

const CartProvider: FC = (props) => {
  const [isOpen, setIsOpen] = useState(false);
  const [products, setProducts] = useState<ICartProduct[]>([]);
  const [total, setTotal] = useState<ICartTotal>(totalInitialValues);

  const CartContextValue: ICartContext = {
    isOpen,
    setIsOpen,
    products,
    setProducts,
    total,
    setTotal,
  };

  return <CartContext.Provider value={CartContextValue} {...props} />;
};

export { CartProvider, useCartContext };
