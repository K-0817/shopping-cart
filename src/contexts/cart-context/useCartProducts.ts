import { useCartContext } from './CartContextProvider';
import useCartTotal from './useCartTotal';
import { ICartProduct } from 'models';

const useCartProducts = () => {
  const { products, setProducts } = useCartContext();
  const { updateCartTotal } = useCartTotal();

  const updateQuantitySafely = (
    currentProduct: ICartProduct,
    targetProduct: ICartProduct,
    quantity: number
  ): ICartProduct => {
    if (currentProduct.id === targetProduct.id) {
      return Object.assign({
        ...currentProduct,
        quantity: currentProduct.quantity + quantity,
      });
    } else {
      return currentProduct;
    }
  };

  const addProduct = (newProduct: ICartProduct) => {
    let updatedProducts;
    const isProductAlreadyInCart = products.some(
      (product: ICartProduct) => newProduct.id === product.id
    );

    if (isProductAlreadyInCart) {
      updatedProducts = products.map((product: ICartProduct) => {
        return updateQuantitySafely(product, newProduct, newProduct.quantity);
      });
    } else {
      updatedProducts = [...products, newProduct];
    }

    setProducts(updatedProducts);
    updateCartTotal(updatedProducts);
  };

  const removeProduct = (productToRemove: ICartProduct) => {
    const updatedProducts = products.filter(
      (product: ICartProduct) => product.id !== productToRemove.id
    );

    setProducts(updatedProducts);
    updateCartTotal(updatedProducts);
  };

  const increaseProductQuantity = (productToIncrease: ICartProduct) => {
    const updatedProducts = products.map((product: ICartProduct) => {
      return updateQuantitySafely(product, productToIncrease, +1);
    });

    setProducts(updatedProducts);
    updateCartTotal(updatedProducts);
  };

  const decreaseProductQuantity = (productToDecrease: ICartProduct) => {
    const updatedProducts = products.map((product: ICartProduct) => {
      return updateQuantitySafely(product, productToDecrease, -1);
    });

    setProducts(updatedProducts);
    updateCartTotal(updatedProducts);
  };

  return {
    products,
    addProduct,
    removeProduct,
    increaseProductQuantity,
    decreaseProductQuantity,
  };
};

export default useCartProducts;
