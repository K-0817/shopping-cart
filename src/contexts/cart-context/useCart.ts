import { useCartContext } from './CartContextProvider';
import useCartProducts from './useCartProducts';
import useCartTotal from './useCartTotal';

const useCart = () => {
  const { isOpen, setIsOpen } = useCartContext();
  const {
    products,
    addProduct,
    removeProduct,
    increaseProductQuantity,
    decreaseProductQuantity,
  } = useCartProducts();
  const { total, updateCartTotal } = useCartTotal();

  const openCart = () => setIsOpen(true);
  const closeCart = () => setIsOpen(false);

  return {
    isOpen,
    openCart,
    closeCart,
    products,
    addProduct,
    removeProduct,
    increaseProductQuantity,
    decreaseProductQuantity,
    total,
    updateCartTotal,
  };
};

export default useCart;
