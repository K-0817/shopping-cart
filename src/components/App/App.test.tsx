import { renderWithThemeProvider } from 'utils/test/test-utils';
import { ProductsProvider } from 'contexts/products-context';
import { CartProvider } from 'contexts/cart-context';

import App from '.';

describe('[components] - App', () => {
  const setup = () => {
    return renderWithThemeProvider(
      <ProductsProvider>
        <CartProvider>
          <App />
        </CartProvider>
      </ProductsProvider>
    );
  };

  test('should render correctly', () => {
    const view = setup();
    expect(view).toMatchSnapshot();
  });
});
