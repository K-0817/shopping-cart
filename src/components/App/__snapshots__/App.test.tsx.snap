// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[components] - App should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App__Container-sc-ebmerl-0"
      >
        <div
          class="Loader__Container-sc-5z440p-0 fnqXlG"
        >
          <div />
          <div />
          <div />
          <div />
        </div>
        <a
          aria-label="View source on Github"
          class="GithubCorner__Container-sc-15lbmbx-1 ilzGAu"
          href="https://github.com/jeffersonRibeiro/react-shopping-cart"
        >
          <svg
            aria-hidden="true"
            height="80"
            style="fill: #151513; color: rgb(255, 255, 255); position: absolute; top: 0px; border: 0px; left: 0px; transform: scale(-1, 1);"
            viewBox="0 0 250 250"
            width="80"
          >
            <path
              d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"
            />
            <path
              class="GithubCorner__OctoArm-sc-15lbmbx-0"
              d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2"
              fill="currentColor"
              style="transform-origin: 130px 106px;"
            />
            <path
              d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z"
              fill="currentColor"
            />
          </svg>
        </a>
        <main
          class="App__TwoColumnGrid-sc-ebmerl-1 buOrsR"
        >
          <div
            class="App__Side-sc-ebmerl-2 dnqiC"
          >
            <div
              class="Filter__Container-sc-bj2vay-0"
            >
              <h4
                class="Filter__Title-sc-bj2vay-2 cVDNWF"
              >
                Sizes:
              </h4>
              <div
                class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
              >
                <label>
                  <input
                    data-testid="checkbox"
                    type="checkbox"
                    value="XS"
                  />
                  <span
                    class="checkmark"
                  >
                    XS
                  </span>
                </label>
              </div>
              <div
                class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
              >
                <label>
                  <input
                    data-testid="checkbox"
                    type="checkbox"
                    value="S"
                  />
                  <span
                    class="checkmark"
                  >
                    S
                  </span>
                </label>
              </div>
              <div
                class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
              >
                <label>
                  <input
                    data-testid="checkbox"
                    type="checkbox"
                    value="M"
                  />
                  <span
                    class="checkmark"
                  >
                    M
                  </span>
                </label>
              </div>
              <div
                class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
              >
                <label>
                  <input
                    data-testid="checkbox"
                    type="checkbox"
                    value="ML"
                  />
                  <span
                    class="checkmark"
                  >
                    ML
                  </span>
                </label>
              </div>
              <div
                class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
              >
                <label>
                  <input
                    data-testid="checkbox"
                    type="checkbox"
                    value="L"
                  />
                  <span
                    class="checkmark"
                  >
                    L
                  </span>
                </label>
              </div>
              <div
                class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
              >
                <label>
                  <input
                    data-testid="checkbox"
                    type="checkbox"
                    value="XL"
                  />
                  <span
                    class="checkmark"
                  >
                    XL
                  </span>
                </label>
              </div>
              <div
                class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
              >
                <label>
                  <input
                    data-testid="checkbox"
                    type="checkbox"
                    value="XXL"
                  />
                  <span
                    class="checkmark"
                  >
                    XXL
                  </span>
                </label>
              </div>
            </div>
            <div
              class="star-button-container"
            >
              <p>
                <small>
                  Leave a star on Github if this repository was useful :)
                </small>
              </p>
              <a
                aria-label="Star jeffersonRibeiro/react-shopping-cart on GitHub"
                class="github-button"
                data-icon="octicon-star"
                data-show-count="true"
                data-size="large"
                href="https://github.com/jeffersonRibeiro/react-shopping-cart"
                tabindex="-1"
              >
                Star
              </a>
            </div>
          </div>
          <main
            class="App__Main-sc-ebmerl-3"
          >
            <main
              class="App__MainHeader-sc-ebmerl-4 jvDqsc"
            >
              <p>
                0
                 Product(s) found
              </p>
            </main>
            <div
              class="Products__Container-sc-uhudcz-0 kTCRJn"
            />
          </main>
        </main>
        <div
          class="Cart__Container-sc-1h98xa9-1 OVjcy"
        >
          <button
            class="Cart__CartButton-sc-1h98xa9-0 Xqoyz"
          >
            <div
              class="Cart__CartIcon-sc-1h98xa9-2 dFOmZq"
            >
              <div
                class="Cart__CartQuantity-sc-1h98xa9-3 gkShyn"
                title="Products in cart quantity"
              >
                0
              </div>
            </div>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App__Container-sc-ebmerl-0"
    >
      <div
        class="Loader__Container-sc-5z440p-0 fnqXlG"
      >
        <div />
        <div />
        <div />
        <div />
      </div>
      <a
        aria-label="View source on Github"
        class="GithubCorner__Container-sc-15lbmbx-1 ilzGAu"
        href="https://github.com/jeffersonRibeiro/react-shopping-cart"
      >
        <svg
          aria-hidden="true"
          height="80"
          style="fill: #151513; color: rgb(255, 255, 255); position: absolute; top: 0px; border: 0px; left: 0px; transform: scale(-1, 1);"
          viewBox="0 0 250 250"
          width="80"
        >
          <path
            d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"
          />
          <path
            class="GithubCorner__OctoArm-sc-15lbmbx-0"
            d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2"
            fill="currentColor"
            style="transform-origin: 130px 106px;"
          />
          <path
            d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z"
            fill="currentColor"
          />
        </svg>
      </a>
      <main
        class="App__TwoColumnGrid-sc-ebmerl-1 buOrsR"
      >
        <div
          class="App__Side-sc-ebmerl-2 dnqiC"
        >
          <div
            class="Filter__Container-sc-bj2vay-0"
          >
            <h4
              class="Filter__Title-sc-bj2vay-2 cVDNWF"
            >
              Sizes:
            </h4>
            <div
              class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
            >
              <label>
                <input
                  data-testid="checkbox"
                  type="checkbox"
                  value="XS"
                />
                <span
                  class="checkmark"
                >
                  XS
                </span>
              </label>
            </div>
            <div
              class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
            >
              <label>
                <input
                  data-testid="checkbox"
                  type="checkbox"
                  value="S"
                />
                <span
                  class="checkmark"
                >
                  S
                </span>
              </label>
            </div>
            <div
              class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
            >
              <label>
                <input
                  data-testid="checkbox"
                  type="checkbox"
                  value="M"
                />
                <span
                  class="checkmark"
                >
                  M
                </span>
              </label>
            </div>
            <div
              class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
            >
              <label>
                <input
                  data-testid="checkbox"
                  type="checkbox"
                  value="ML"
                />
                <span
                  class="checkmark"
                >
                  ML
                </span>
              </label>
            </div>
            <div
              class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
            >
              <label>
                <input
                  data-testid="checkbox"
                  type="checkbox"
                  value="L"
                />
                <span
                  class="checkmark"
                >
                  L
                </span>
              </label>
            </div>
            <div
              class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
            >
              <label>
                <input
                  data-testid="checkbox"
                  type="checkbox"
                  value="XL"
                />
                <span
                  class="checkmark"
                >
                  XL
                </span>
              </label>
            </div>
            <div
              class="Filter__Checkbox-sc-bj2vay-1 cUzHAu"
            >
              <label>
                <input
                  data-testid="checkbox"
                  type="checkbox"
                  value="XXL"
                />
                <span
                  class="checkmark"
                >
                  XXL
                </span>
              </label>
            </div>
          </div>
          <div
            class="star-button-container"
          >
            <p>
              <small>
                Leave a star on Github if this repository was useful :)
              </small>
            </p>
            <a
              aria-label="Star jeffersonRibeiro/react-shopping-cart on GitHub"
              class="github-button"
              data-icon="octicon-star"
              data-show-count="true"
              data-size="large"
              href="https://github.com/jeffersonRibeiro/react-shopping-cart"
              tabindex="-1"
            >
              Star
            </a>
          </div>
        </div>
        <main
          class="App__Main-sc-ebmerl-3"
        >
          <main
            class="App__MainHeader-sc-ebmerl-4 jvDqsc"
          >
            <p>
              0
               Product(s) found
            </p>
          </main>
          <div
            class="Products__Container-sc-uhudcz-0 kTCRJn"
          />
        </main>
      </main>
      <div
        class="Cart__Container-sc-1h98xa9-1 OVjcy"
      >
        <button
          class="Cart__CartButton-sc-1h98xa9-0 Xqoyz"
        >
          <div
            class="Cart__CartIcon-sc-1h98xa9-2 dFOmZq"
          >
            <div
              class="Cart__CartQuantity-sc-1h98xa9-3 gkShyn"
              title="Products in cart quantity"
            >
              0
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
